backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/discussion/DiscussionService.java

```diff
@@ -0,0 +1,273 @@
+import pt.ulisboa.tecnico.socialsoftware.tutor.answer.domain.QuestionAnswer;
+import pt.ulisboa.tecnico.socialsoftware.tutor.answer.domain.QuizAnswer;
+import pt.ulisboa.tecnico.socialsoftware.tutor.answer.repository.QuizAnswerRepository;
+import pt.ulisboa.tecnico.socialsoftware.tutor.discussion.domain.Discussion;
+import pt.ulisboa.tecnico.socialsoftware.tutor.discussion.domain.Reply;
+import pt.ulisboa.tecnico.socialsoftware.tutor.discussion.dto.DiscussionDto;
+import pt.ulisboa.tecnico.socialsoftware.tutor.discussion.dto.ReplyDto;
+import pt.ulisboa.tecnico.socialsoftware.tutor.discussion.repository.DiscussionRepository;
+import pt.ulisboa.tecnico.socialsoftware.tutor.discussion.repository.ReplyRepository;
+import pt.ulisboa.tecnico.socialsoftware.tutor.exceptions.TutorException;
+import pt.ulisboa.tecnico.socialsoftware.tutor.question.api.TopicController;
+import pt.ulisboa.tecnico.socialsoftware.tutor.question.domain.Question;
+import pt.ulisboa.tecnico.socialsoftware.tutor.question.repository.QuestionRepository;
+import pt.ulisboa.tecnico.socialsoftware.tutor.user.domain.User;
+import pt.ulisboa.tecnico.socialsoftware.tutor.user.repository.UserRepository;
+
+import javax.persistence.EntityManager;
+import javax.persistence.PersistenceContext;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static pt.ulisboa.tecnico.socialsoftware.tutor.exceptions.ErrorMessage.*;
+
+@Service
+public class DiscussionService {
+    @Autowired
+    public DiscussionRepository discussionRepository;
+
+    @Autowired
+    public UserRepository userRepository;
+
+    @Autowired
+    public QuestionRepository questionRepository;
+
+    @Autowired
+    public ReplyRepository replyRepository;
+
+    @Autowired
+    public QuizAnswerRepository quizAnswerRepository;
+
+    @PersistenceContext
+    EntityManager entityManager;
+
+    @Retryable(
+            value = { SQLException.class },
+            backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.READ_COMMITTED)
+    public DiscussionDto createDiscussion(int quizAnswerId, int questionId, DiscussionDto discussionDto) {
+        checkDiscussionDto(discussionDto);
+        User user = userRepository.findById(discussionDto.getUserId()).orElseThrow(() -> new TutorException(USER_NOT_FOUND, discussionDto.getUserId()));
+        Question question = questionRepository.findById(discussionDto.getQuestion().getId()).orElseThrow(() -> new TutorException(QUESTION_NOT_FOUND, discussionDto.getQuestion().getId()));
+        QuizAnswer quizAnswer = quizAnswerRepository.findById(quizAnswerId).orElse(null);
+
+        checkUserAndQuestion(user, question);
+        checkMessage(discussionDto.getMessage());
+        checkQuizAnswer(quizAnswer);
+
+
+        List<QuestionAnswer> questionAnswerList = quizAnswer.getQuestionAnswers().stream().filter(qAnswer -> {
```

Comment:
I think that this, till line 83 can be removed because waht is being checked is whehter the same question appers twice in a quiz, which should not be the concern of this code